### PR TITLE
(BIDS-2258) use chart_series for some consensus-charts aswell

### PR DESCRIPTION
**Pull request by guybrush**
### <samp>🤖 Generated by Copilot at f7788c4</samp>

This pull request enhances the chart series functionality for chains with execution layer, improves the performance and consistency of the withdrawals and deposits charts, and refactors and fixes some minor issues in the code. It affects the files `cmd/statistics/main.go`, `db/statistics.go`, `services/charts_updater.go`, and `db/db.go`.

### CODE REVIEW

File: db/migrations/20230629144437_add_table_graffiti_stats.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
+-- +goose Up
```
Comment (recy21) [inline db/migrations/20230629144437_add_table_graffiti_stats.sql:1]: Create / Remove always with `CONCURRENTLY`

File: db/statistics.go [hunk 2]
```diff
@@ -1260,10 +1260,118 @@ func GetValidatorIncomeHistory(validatorIndices []uint64, lowerBoundDay uint64,
 
 func WriteChartSeriesForDay(day int64) error {
 	startTs := time.Now()
+	err := WriteExecutionChartSeriesForDay(day)
```
Comment (recy21) [inline db/statistics.go:1263]: Would it make sense to start a go routine for this? (and WriteConsensusChartSeriesForDay)

File: db/statistics.go [hunk 3]
```diff
@@ -1526,10 +1634,14 @@ func WriteChartSeriesForDay(day int64) error {
...
 	}
 
-	logger.Infof("Exporting MARKET_CAP: %v", newEmission.Div(decimal.NewFromInt(1e18)).Add(decimal.NewFromFloat(72009990.50)).Mul(decimal.NewFromFloat(price.GetEthPrice("USD"))).String())
-	err = SaveChartSeriesPoint(dateTrunc, "MARKET_CAP", newEmission.Div(decimal.NewFromInt(1e18)).Add(decimal.NewFromFloat(72009990.50)).Mul(decimal.NewFromFloat(price.GetEthPrice("USD"))).String())
-	if err != nil {
-		return fmt.Errorf("error calculating MARKET_CAP chart_series: %w", err)
+	switch utils.Config.Chain.Config.DepositChainID {
+	case 1:
+		crowdSale := 72009990.50
+		logger.Infof("Exporting MARKET_CAP: %v", newEmission.Div(decimal.NewFromInt(1e18)).Add(decimal.NewFromFloat(crowdSale)).Mul(decimal.NewFromFloat(price.GetEthPrice("USD"))).String())
```
Comment (recy21) [inline db/statistics.go:1640]: As used twice, we should use a variable for `newEmission.Div(decimal.NewFromInt(1e18)).Add(decimal.NewFromFloat(crowdSale)).Mul(decimal.NewFromFloat(price.GetEthPrice("USD"))).String()`

File: services/charts_updater.go [hunk 4]
```diff
@@ -321,24 +312,26 @@ func stakedEtherChartData() (*types.GenericChartData, error) {
...
-		Epoch         uint64
-		EligibleEther uint64
+	data := []struct {
+		Indicator string    `db:"indicator"`
+		Time      time.Time `db:"time"`
+		Value     float64   `db:"value"`
 	}{}
 
-	err := db.ReaderDb.Select(&rows, "SELECT epoch, eligibleether FROM epochs ORDER BY epoch")
+	epoch := LatestEpoch()
```
Comment (recy21) [inline services/charts_updater.go:321]: Suggestion: Create a function which return LatestEpoch - 1
(easier to read)

File: services/charts_updater.go [hunk 5]
```diff
@@ -801,67 +793,41 @@ func performanceDistribution365dChartData() (*types.GenericChartData, error) {
...
-	err = db.ReaderDb.Select(&eth1Rows, `
-		select
-			extract(epoch from block_ts)::int as timestamp,
-			amount,
-			valid_signature as valid
-		from eth1_deposits
-		order by timestamp`)
-	if err != nil {
-		return nil, fmt.Errorf("error getting eth1-deposits: %w", err)
+	if LatestEpoch() == 0 {
```
Comment (recy21) [inline services/charts_updater.go:796]: This is strange, as a few lines below, we check again and reduce by one. We shouldn't call the function twice and we should always the same value / calculation.

Comment (recy21) [reply services/charts_updater.go:796]: Same situation appears again.

**Review by recy21 — CHANGES_REQUESTED** [review]
Did a quick review, didn't check the queries. If a full review required, it will take more time, a lot of complex changes.

File: db/migrations/20230629144437_add_table_graffiti_stats.sql (hunk 1 above)
Comment (guybrush) [reply db/migrations/20230629144437_add_table_graffiti_stats.sql:1]: oriented here: https://github.com/gobitfly/eth2-beaconchain-explorer/blob/master/db/migrations/20230614122901_add_ens_tables.sql

the problem with using concurrently here is that I need to split the file up in transactions and non-transaction-queries (will do)

File: db/statistics.go (hunk 3 above)
Comment (guybrush) [reply db/statistics.go:1640]: its the sane for all variables throughout the whole function

eg: 
- https://github.com/gobitfly/eth2-beaconchain-explorer/blob/21451456af9d9661be2f06100324764047b7b7dc/db/statistics.go#L1630
- https://github.com/gobitfly/eth2-beaconchain-explorer/blob/21451456af9d9661be2f06100324764047b7b7dc/db/statistics.go#L1622
- etc

i think we should be consistent here

File: services/charts_updater.go (hunk 4 above)
Comment (guybrush) [reply services/charts_updater.go:321]: this is copy paste from all other places in this file, wanted to be consistent

if we use `LatestEpoch() - 1` we still have to check if `LatestEpoch() > 0`

File: db/migrations/20230629144437_add_table_graffiti_stats.sql (hunk 1 above)
Comment (guybrush) [reply db/migrations/20230629144437_add_table_graffiti_stats.sql]: fixed

File: db/migrations/20230629144437_add_table_graffiti_stats.sql [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+-- +goose StatementBegin
+SELECT 'up SQL query - remove graffiti_stats tables';
```
[the remark this answers is not in the document]
Comment (guybrush) [reply db/migrations/20230629144437_add_table_graffiti_stats.sql]: fixed

File: db/migrations/20230629144437_add_table_graffiti_stats.sql [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        proposer_count INTEGER NOT NULL,
+        PRIMARY KEY (graffiti, day)
+    );
+CREATE INDEX IF NOT EXISTS idx_graffiti_stats_day ON graffiti_stats (day);
+CREATE INDEX IF NOT EXISTS idx_graffiti_stats_day_graffiti_text ON graffiti_stats USING gin (graffiti_text gin_trgm_ops);
+-- +goose StatementEnd
+
+-- +goose Down
+-- +goose StatementBegin
+SELECT 'down SQL query - remove ens lookup tables';
```
[the remark this answers is not in the document]
Comment (guybrush) [reply db/migrations/20230629144437_add_table_graffiti_stats.sql]: fixed

File: services/charts_updater.go (hunk 4 above)
Comment (guybrush) [reply services/charts_updater.go]: fixed

File: services/charts_updater.go (hunk 5 above)
Comment (guybrush) [reply services/charts_updater.go:766]: fixed

File: db/statistics.go (hunk 2 above)
Comment (guybrush) [reply db/statistics.go]: good idea, fixed

**Review by recy21 — APPROVED** [review]

**Pull request merged**